pkg/provisioningv2/rke2/planner/planner.go

```diff
@@ -883,35 +886,22 @@ func PruneEmpty(config map[string]interface{}) {
 	}
 }
 
-func addAddresses(secrets corecontrollers.SecretCache, config map[string]interface{}, machine *capi.Machine) {
-	if data := machine.Annotations[InternalAddressAnnotation]; data != "" {
-		config["node-ip"] = data
-	}
-
-	if data := machine.Annotations[AddressAnnotation]; data != "" {
-		config["node-external-ip"] = data
-		// Don't check for an IP address from the secret if it is already provided.
-		return
-	}
+func addAddresses(secrets corecontrollers.SecretCache, controlPlaneIPs []interface{}, config map[string]interface{}, machine *capi.Machine) {
+	internalIPAddress, ipAddress := getIPAddresses(secrets, machine)
+	setNodeExternalIP := ipAddress != "" && internalIPAddress != "" && ipAddress != internalIPAddress
 
-	if config["cloud-provider-name"] != nil || machine.Annotations["objectset.rio.cattle.io/id"] == "unmanaged-machine" {
-		// Unmanaged machines won't have a state secret,
-		// and the cloud provider will handle this, if it is set.
-		return
+	if setNodeExternalIP && !isOnlyWorker(machine) {
+		config["advertise-address"] = internalIPAddress
+		appendToSlice(config, "tls-san", controlPlaneIPs...)
 	}
 
-	secret, err := secrets.Get(machine.Spec.InfrastructureRef.Namespace, name.SafeConcatName(machine.Spec.InfrastructureRef.Name, "machine", "state"))
-	if err == nil && secret != nil && len(secret.Data["extractedConfig"]) != 0 {
-		driverConfig, err := nodeconfig.ExtractConfigJSON(base64.StdEncoding.EncodeToString(secret.Data["extractedConfig"]))
-
-		if err == nil && len(driverConfig) != 0 {
-			ipAddress := convert.ToString(values.GetValueN(driverConfig, "Driver", "IPAddress"))
-			internalIPAddress := convert.ToString(values.GetValueN(driverConfig, "Driver", "PrivateIPAddress"))
+	if config["cloud-provider-name"] == nil {
```

Comment:
You need to check for nil and `""` because `""` value actually means unset.
```suggestion
	if convert.ToString(config["cloud-provider-name"]) == "" {
```

Reply:
Done. Also checked for empty strings in the annotations when provided.